### PR TITLE
[2.13.x] DDF-4457 Encrypt file system storage

### PR DESCRIPTION
**Pull request by bakejeyner**
#### What does this PR do?
Updates the `FileSystemStorageProvider` to encrypt the content it writes to disk.

#### Who is reviewing it? 
@blen-desta 
@garrettfreibott 
@stustison 

#### Select relevant component teams: 
@codice/security 

#### How should this be tested?
1. Build DDF and install the standard profile.
1. Ingest a resource using Intrigue.
1. Query for that resource using Intrigue.
1. Verify that the result looks normal
    1. Check for a thumbnail (if applicable)
    1. Verify all of the `Actions` in the `Inspector` tab work, especially the `resource` action.
1. Verify that the content is encrypted on disk
   1. Navigate to `<DDF_HOME>/data/content/store/{first 3 characters of id}/{next three characters of id}/{id}`, and verify that you cannot open the resource.
1. Verify that the new keysets have been generated.
    1. Navigate to the keyset directory under `<DDF_HOME>/etc/keysets`
    1. Verify that there are 6 total entries: `encryption-persistence-manager.json`, `encryption-persistence-manager-streaming.json`, `encrypiton-service.json`, `encrypiton-service-streaming.json`, `file-system.json`, and `file-system-streaming.json`

#### What are the relevant tickets?
[DDF-4457](https://codice.atlassian.net/browse/DDF-4457)

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-localstorageprovider/pom.xml [hunk 1]
```diff
@@ -63,6 +63,11 @@
             <groupId>joda-time</groupId>
             <artifactId>joda-time</artifactId>
         </dependency>
+      <dependency>
+        <groupId>ddf.security.encryption</groupId>
+        <artifactId>security-encryption-crypter</artifactId>
+        <version>2.13.7-SNAPSHOT</version>
```
Comment (bdeining) [inline catalog/core/catalog-core-localstorageprovider/pom.xml]: `${project.version}`

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java [hunk 2]
```diff
@@ -162,10 +167,16 @@ synchronized String encrypt(String plainTextValue) {
...
+        LOGGER.debug(
+            "Failed to encrypt value of %s, because it was null or blank.", plainTextValue);
+        return plainTextValue;
+      }
+
       try {
         return crypter.encrypt(plainTextValue);
       } catch (CrypterException e) {
-        LOGGER.warn(String.format("Failed to encrypt to bundle cache. %s", e.getCause()));
+        LOGGER.warn(String.format("Failed to encrypt to bundle cache. %s", e.getMessage()));
```
Comment (bdeining) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java]: this should be parameterized

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java [hunk 3]
```diff
@@ -162,10 +167,16 @@ synchronized String encrypt(String plainTextValue) {
       if (initFailureMessage != null) {
         alertSystem(initFailureMessage, initException);
       }
+      if (isBlank(plainTextValue)) {
+        LOGGER.debug(
+            "Failed to encrypt value of %s, because it was null or blank.", plainTextValue);
```
Comment (bdeining) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java]: should this be parameterized?

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java [hunk 4]
```diff
@@ -175,10 +186,16 @@ synchronized String decrypt(String encryptedValue) {
       if (initFailureMessage != null) {
         alertSystem(initFailureMessage, initException);
       }
+      if (isBlank(encryptedValue)) {
+        LOGGER.debug(
+            "Failed to decrypt value of %s, because it was null or blank.", encryptedValue);
```
Comment (bdeining) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java]: should this be parameterized?

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java [hunk 5]
```diff
@@ -175,10 +186,16 @@ synchronized String decrypt(String encryptedValue) {
...
+        return encryptedValue;
+      }
+
       try {
         return crypter.decrypt(encryptedValue);
       } catch (CrypterException e) {
-        LOGGER.warn(String.format("Failed to decrypt from bundle cache. %s", e.getCause()));
+        LOGGER.warn(String.format("Failed to decrypt from bundle cache. %s", e.getMessage()));
```
Comment (bdeining) [inline platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java]: should this be parameterized?

File: platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java (hunk 2 above)
Comment (bakejeyner) [reply platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java]: Forgot I could use `{}` in the Logger. Changing now 🎉

**Comment by jlcsmith** [comment]
Note - we should probably verify that metacards with derived content still work.  May require testing in Alliance.

**Comment by bakejeyner** [comment]
build now

File: catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java [hunk 6]
```diff
@@ -562,17 +582,17 @@ private ContentItem generateContentFile(
...
-    long copy;
-
     if (storeReference != null) {
-      copy = item.getSize();
+      copySize = item.getSize();
+      String encryptedReference = crypter.encrypt(storeReference);
       Files.write(
           Paths.get(contentItemPath.toString() + "." + REF_EXT),
-          storeReference.getBytes(Charset.forName("UTF-8")));
+          encryptedReference.getBytes(Charset.defaultCharset()));
```
Comment (stustison) [inline catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java]: ✏️ I don't think we want to rely on the default charset here. This is going to be different depending on the OS, that is why we are using UTF-8 all over the place.

**Comment by bakejeyner** [comment]
On an Alliance with this branch's changes, I ingested `withbe.ntf`. I verified that the `derived-download-uri`s work properly. I was able to see the `Overview` (thumbnail looking thing) and the `original`.

However, the `original` is not equal to the original ntif. Not sure if this is expected or not.

**Comment by jlcsmith** [comment]
@bakejeyner that is the correct behavior.  Thanks for verifying that

**Comment by bakejeyner** [comment]
build now
(fixed the CDM component tests, having trouble reproducing the Itest failure locally)

**Comment by bakejeyner** [comment]
build now
(fixed feature dependency)

**Comment by bakejeyner** [comment]
build now
(🤞)

**Comment by bakejeyner** [comment]
build now

**Comment by bakejeyner** [comment]
build now
(typo 🤦‍♂️)

File: catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java [hunk 7]
```diff
@@ -388,88 +401,137 @@ public void rollback(StorageRequest request) throws StorageException {
...
-        throw new StorageException(e);
+    String filename = path.getFileName().toString();
+
+    // resolve external reference if necessary, determine the extension, and retrieve an
+    // InputStream to the content
+    InputStream contentInputStream;
+    String extension;
+
+    try {
+      if (REF_EXT.equals(FilenameUtils.getExtension(path.getFileName().toString()))) {
```
Comment (coyotesqrl) [inline catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java]: ✏️ Would clarify and simplify the code in this block if you used `filename` as defined on line 411 above.

File: catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java [hunk 8]
```diff
@@ -388,88 +401,137 @@ public void rollback(StorageRequest request) throws StorageException {
...
+        // remove the external reference extension so we can get the real extension
+        extension =
+            FilenameUtils.getExtension(
+                FilenameUtils.removeExtension(path.getFileName().toString()));
+        contentInputStream = getInputStreamFromReference(path);
+      } else {
+        extension = FilenameUtils.getExtension(path.getFileName().toString());
+        contentInputStream = getInputStreamFromResource(path);
```
Comment (coyotesqrl) [inline catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java]: ✏️ This and line 424 in the `if` block are identical. Can be pulled out to right below the `if/else` at the end of the `try` block.

File: features/security/src/main/feature/feature.xml [hunk 9]
```diff
@@ -624,4 +624,13 @@
              description="Verifies that the installation is located in a secure directory.">
         <bundle>mvn:ddf.platform.security/secure-boot/${project.version}</bundle>
     </feature>
+
+    <feature name="tink-test-dependencies" version="${project.version}"
```
Comment (coyotesqrl) [inline features/security/src/main/feature/feature.xml]: ❓ It seems odd to have a test dependencies feature here. Do we do this anywhere else?

File: catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java (hunk 8 above)
Comment (bakejeyner) [reply catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java]: Line 424 is
```
        contentInputStream = getInputStreamFrom**Reference**(path);
```
whereas line 427 is
```
        contentInputStream = getInputStreamFrom**Resource**(path);
```

File: features/security/src/main/feature/feature.xml (hunk 9 above)
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml]: We only need this feature for component/integration tests that need to load the tink dependencies. Let me 👀 if there is a better place to put this real quick.

File: catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java (hunk 8 above)
Comment (coyotesqrl) [reply catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java]: Man. My eyes crossed on that. Thanks.

File: features/security/src/main/feature/feature.xml (hunk 9 above)
Comment (bakejeyner) [reply features/security/src/main/feature/feature.xml]: Decided to throw the feature in `<ddf_source>\features\test-utilities\src\main\feature\feature.xml` instead.

File: catalog/core/catalog-core-localstorageprovider/src/test/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProviderTest.java [hunk 10]
```diff
@@ -122,7 +132,8 @@ public void setUp() throws IOException {
 
   @After
   public void clean() {
-    tempFolder.delete();
+    System.clearProperty(SecurityConstants.KEYSET_DIR);
```
Comment (blen-desta) [inline catalog/core/catalog-core-localstorageprovider/src/test/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProviderTest.java]: ❓ Do we not need that deletion?

Comment (bakejeyner) [reply catalog/core/catalog-core-localstorageprovider/src/test/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProviderTest.java]: No, the temp folder is automatically created/deleted per test method when the `@Rule` annotation is specified at it's initialization:
```
@Rule public TemporaryFolder temporaryFolder = new TemporaryFolder();
```

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/DeleteOperations.java [hunk 11]
```diff
@@ -179,7 +179,7 @@ public DeleteResponse doDelete(DeleteRequest deleteRequest, List<String> fanoutT
 
     } catch (RuntimeException re) {
       LOGGER.info("Exception during runtime while performing delete", re);
-      throw new InternalIngestException("Exception during runtime while performing delete");
+      throw new InternalIngestException("Exception during runtime while performing delete.", re);
```
Comment (stustison) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/operations/DeleteOperations.java]: ❗️ Shouldn't log and rethrow. You can put this back the way it was.

**Comment by bakejeyner** [comment]
build now

**Comment by tyler30clemens** [comment]
Hero Successful ✅ 
verified all the steps work

**Pull request merged**